Travis test speedups

chrisr-diffblue opened the pull request:
This PR does two things to help speed up or Travis CI jobs. Firstly it makes the Makefile based jobs run the regression tests in two parallel jobs, using GNU Parallel to mimic the behaviour that the CMake based jobs currently have. Secondly, it tweaks the ordering of the list of regression test suite directories to enable us to get better job scheduling and therefore better utilisation of two parallel jobs.

chrisr-diffblue committed:
Enable a coarse-grained prallelism when running regression test jobs via Makefiles

Previously, the makefiles ran the regression test target sequentially,
regardless of any -jN option passed to make. In this commit, an extra
'test-batch' target is added which enables running the same regression
tests, but with each top-level test directory run as a seperate target.
To avoid horribly interleaved output from parallel test jobs, GNU parallel
is used to execute the test jobs. This provides line bufferering, such that
output from each job is prefixed with a job number. This matches the behaviour
of ctest used in the CMake based builds.

thk123 commented on regression/Makefile:
```diff
@@ -20,18 +20,51 @@ DIRS = ansi-c \
+
+# Run a single test directory, capturing it's stdout/stderr to
+# a file
+.%-batch-output:
```
Can/are equivalent changes done for the CMake version?

chrisr-diffblue replied:
The CMake version uses CTest to run the tests, at that already knows how to do this (and does it more elegantly, to be honest). This is just makeing (no pun intended) the Makefile builds catchup.

thk123 commented on regression/Makefile:
```diff
@@ -20,18 +20,51 @@ DIRS = ansi-c \
+.%-batch-output:
+	@if [ -d "$*" ]; then \
+		echo "Running $*..." ; \
+		$(MAKE) "test-$*" > "$@" 2>&1 || (echo "Failed $* at $$(date)" ; cat "$@" ; exit 1) ; \
```
No output if the tests all pass right? Seems reasonable mostly just noting that is a change in behaviour.

chrisr-diffblue replied:
Currently the `test-batch` rule will output all the output if no tests fail, so you'll still get the full output, just batched up to the end of test runs. This is all still a work in progress though, which is why I have the 'DO NOT MERGE' label on this PR at the moment. I want to play with a few different strategies first rather than knee jerk something in :-)

thk123 reviewed (approved):
Seems fine and seems like a big speedup? Maybe just getting more lucky with caching but just checked my newest PR and seems to take twice as long on the same stage!

chrisr-diffblue commented:
I think probably just luck with caching/host load, but I've seen 5-10 minutes being shaved off the test run times, but Travis is so badly inconsistent with execution times I'd hesitate to claim more than that with any confidence. I do believe there is still some savings to be squeezed out of this approach yet though, namely around being a bit smarter about scheduling the test directories given the un-even distribution of runtime/size of each directory. I'm going to try that out this morning.

tautschnig commented:
Wouldn't it be possible to use native `make` support for proper interleaving of output: https://www.gnu.org/software/make/manual/html_node/Parallel-Output.html#Parallel-Output ?

chrisr-diffblue commented:
@tautschnig We currently use make 3.81, otherwise yes, that's what I would have used :-) If people are happy to bump up the minimum version of make, then I'm happy to make that change.

tautschnig commented:
Err, the terrible world of legacy software :-)   So how about a somewhat different proposal, and I'd like to highlight this thread to @karkhaz: Leave the absence of concurrency at test-suite (aka directory) level alone, but instead focus on parallelism within each directory? That would also resolve the problem of vastly different runtimes.

@zemanlx had looked into this before and test.pl already has support for parallelism, except that it apparently doesn't really work or doesn't work any longer. @karkhaz and I discussed several options last week; directly using the support for parallelism in the build tools would seem preferable, but we'd then run into the same limitations of make <= 4.0. Hence sorting it out within test.pl might be the best route forward. My preference would be a rewrite of this historic piece of software in Python.

chrisr-diffblue commented:
Thanks for the thoughts @tautschnig - much appreciated. I agree that the experiments in this PR are not the long term solution, and I'd very much like to come up with a proper solution along the lines you say (i.e. re-write test.pl if needs be, but ideally be able to use the build system) but I was just trying to see if there was any low-hanging fruit we could pick without anything too disgusting in the short term.

chrisr-diffblue committed:
Improve job scheduling when running regression tests with -jN

Currently the regression test suite directories are listed and executed in
alphabetical order. However, because the runtimes of the various test suites
are not uniform, this leads to poor utilisation of the parallel jobs. For
instance, consider four tests, A, B, C and D. Lets assume the runtimes for these
hypothetical tests are:

  A 100
  B 50
  C 50
  D 100

further, lets assume we have two cores to run on. The tests are run
in alphabetical order, leading to a job schedule that looks something like:

Time  Core     Test
0     1        A
0     2        B
50    2        C
100   1        D
100   2        <idle>
200 - All tests complete

However, if jobs are scheduled in roughly decending runtime order, we get a
job schedule that looks roughly like:

Time  Core     Test
0     1        A
0     2        D
100   1        B
100   2        C
150 - All tests complete

This commit re-orders the list of regression suite directories into roughly
decending runtime order.

karkhaz commented:
@tautschnig thanks for the ping! I paused work on this because I wanted to get the discussion going on #1915. I'll resume work on this now while we flesh out details of #1915. I'll proceed with writing `test.pl` in Python if that makes it easier to work with old versions of make, and let you know how that goes. I'd be happy to hear any insights that you glean from your experiments @chrisr-diffblue!

chrisr-diffblue commented:
@karkhaz That's great to hear. There's a few people here inside Diffblue who are also interested in updating our test infrastructure, and who have some experience using PyTest as a harness, so they may be interested in collaborating and exchange ideas. @owen-jones-diffblue would definitely be interested. 

In terms of my experiments in this PR, this is just meant to be a quick fix to gain us some better utilisation of the two cores we have available on every travis host. This seems to have enabled us to get a full CI run in somewhere around 35-40 minutes depending on how lucky we get with travis load, but it won't scale nicely with number of cores (i.e. if you try a -j8 or something) because of the lack of test-level parallelism.

chrisr-diffblue commented:
@tautschnig I've tweaked this patch to the Makefiles to now use GNU Parallel instead and this now pretty much does exactly what the existing CMake based build/tests do - would that satisfy you as a short term fix until we have a longer term re-write of the test system in place? With these two commits, I've been able to get full Travis CI jobs completing in 35-40 minutes which seems a worthwhile improvement for minimal extra ugliness... but I realise this could be controversial :-) Just to also be clear for the avoidance of doubt, the 'solution/hack' in this PR won't really scale that well beyond 2 cores, so its hopefully not going to take away motivation to fix this stuff properly either.

owen-mc-diffblue commented:
Indeed, I'm a big fan of pytest. My team have been using it (in a private repo) and it suits us very well. We haven't tried parallelising the tests, but I just did a bit of googling and it seems like there is a plug-in called xdist that enables this. If people agree that using an existing framework like pytest is a viable alternative to rewriting test.pl in python then I could try pytest-xdist out and check it does what we want.

I've also thought a bit about how to run the existing tests using pytest. It wouldn't be hard to write a function `run_test_desc` to interpret existing `test.desc` files. So for each existing test.pl test we'd make a small function (tests are functions in pytest) which calls `run_test_desc` with the path to the test.desc file. There would have to be some minor tweaks to account for different `chain.sh` behaviours. Tags for a given test would be added as decorators to the corresponding function. That would allow us to support all the existing tests without too much effort, and new tests could still be written in the test.desc manner or using pytest directly. This would allow for more involved parsing of the output (e.g. checking that whatever occurs in the this location in one line occurs in this other location in the next line) and more sharing of functionality between tests.

tautschnig commented on regression/Makefile:
```diff
@@ -20,15 +20,37 @@ DIRS = ansi-c \
+# Run all test directories in sequence
+.PHONY: test
 test:
 	@for dir in $(DIRS); do \
```
How  about `@for dir in $(wildcard $(DIRS)); do` to have make filter out non-existent directories?

chrisr-diffblue replied:
Thanks for that suggestion, I can make things a lot simpler with that idea.

tautschnig commented on regression/Makefile:
```diff
@@ -20,15 +20,37 @@ DIRS = ansi-c \
+	if [ -d "$*" ]; then \
+		echo "Running $*..." ; \
+		$(MAKE) -C "$*" test || exit 1; \
+	fi;
```
See above for how to avoid the extra test here, at which point this rule could be simplified?

tautschnig commented on regression/Makefile:
```diff
@@ -20,15 +20,37 @@ DIRS = ansi-c \
+		--linebuffer \
+ 		--jobs $(JOBS) \
+ 		$(MAKE) "test-{}" \
+ 			::: $(DIRS)
```
I'm not sure what the indentation tells here? Actually different pages on GitHub show it in different ways, looks like a mix of space and tabs.

chrisr-diffblue replied:
It tells that @chrisr-diffblue can't type... :-)

tautschnig commented:
I've been using `pytest` with some Python projects and surely it makes a lot of sense for those (but I am by no means an expert). What is less clear to me, though, is the value it would bring to _this_ project. As @owen-jones-diffblue described, we would need to add some wrappers and parsers, as is the case for a potential rewrite of `test.pl` in Python. What are the remaining features that `pytest` would natively bring to the table?

thk123 commented:
@tautschnig I think vs. a rewrite of `test.pl` in python, pytest would at least save us writing:
- the command line interface
- support for tagging tests
- support for rerunning failed tests
- IDE and other tool integrations
- support for parallel running (via `xdist`)??
- easy extension of tests to be not just line by line regex matchers down the road
- documentation for how to run the testing framework(!)

Of course most of these fall into the "nice to have" bucket, but I wonder what a home grown rewrite of `test.pl` brings over `pytest`.

chrisr-diffblue commented:
@tautschnig I've now cleaned up the Makefile change in line with your suggestion to use `$(wildcard)` - many thanks.

chrisr-diffblue commented:
@tautschnig I know you're busy, so apologies for the ping, but any chance of a quick re-review of this?

tautschnig commented:
Just to respond to @thk123's very helpful explanation: my slight resistance against `pytest` is just the general resistance against dependencies as each of them as each of them may break or may be unavailable on some platform. From the list of features, however, to me it clearly seems beneficial use `pytest`.
